keps/sig-apps/2804-consolidate-workload-controllers-status/README.md

```diff
@@ -0,0 +1,808 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As an end-user of Kubernetes, I'd like all my workload controllers to have consistent states
+
+#### Story 2
+As an developer building Kubernetes Operators, I'd like all my operators deployed to have
+consistent states.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+We are proposing 3 new conditions called
+- Progressing
+- Complete
+- Failed
+to be added to DaemonSet and Stateful controllers. The definitions are similar to what we have for [Deployment controller](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#deployment-status).
+
+#### Progressing
+Kubernetes marks a DaemonSet or Stateful as `progressing` when:
+- The DaemonSet or StatefulSet is created
+- The DaemonSet or StatefulSet is scaling up or scaling down
+- New DaemonSet or StatefulSet pods become Ready or available
+
+#### Complete
+Kubernetes marks a DaemonSet or Stateful as complete when it has the following characteristics:
+
+- All of the replicas associated with the DaemonSet or StatefulSet have been updated to the latest version you've specified, meaning any updates you've requested have been completed.
+- All of the replicas associated with the DaemonSet or StatefulSet are available.
+- No old replicas for the DaemonSet or Stateful are running.
+
+#### Failed
+In order to introduce this condition we need to come up with a new field called `.spec.progressDeadlineSeconds` which denotes the number of seconds the controller waits before indicating(in the workload controller status) that the controller progress has stalled.
+
+There are many factors that can cause failure to happen like:
+- Insufficient quota
+- Readiness probe failures
+- Image pull errors
+
+
+See the [Kubernetes API Conventions](https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) for more information on status conditions
+
+Because of the number of changes that are involved as part of this effort, we are thinking of a phased approach where we introduce these conditions to DaemonSet controller behind a featuregate flag first in one release and then make similar changes to StatefulSet controller. This also needs some code refactoring of existing conditions for Deployment controller.
```

Comment:
wrt to Deployments, are we going to copy ReplicaFailure values to a new Failed condition for consistency?